packages/medusa/src/services/payment-collection.ts

```diff
@@ -160,4 +189,493 @@ export default class PaymentCollectionService extends TransactionBaseService {
+          MedusaError.Types.NOT_ALLOWED,
+          `Payment ${payment.id} is not captured`
+        )
+      }
+
+      const refundable = payment.amount - payment.amount_refunded
+      if (amount > refundable) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Only ${refundable} can be refunded from Payment ${payment.id}`
+        )
+      }
+
+      let refundError: Error | null = null
+      const refund = await this.paymentProviderService_
+        .withTransaction(manager)
+        .refundFromPayment(payment, amount, reason, note)
+        .catch((err) => {
+          refundError = err
+        })
+
+      payCol.refunded_amount = payCol.refunded_amount ?? 0
+      if (refund) {
+        payCol.refunded_amount += refund.amount
+      }
+
+      if (payCol.refunded_amount === 0) {
+        payCol.status = PaymentCollectionStatus.REQUIRES_ACTION
+      } else if (payCol.refunded_amount === payCol.amount) {
+        payCol.status = PaymentCollectionStatus.REFUNDED
+      } else {
+        payCol.status = PaymentCollectionStatus.PARTIALLY_REFUNDED
+      }
+
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      await paymentCollectionRepository.save(payCol)
+
+      if (!refund) {
+        await this.eventBusService_
+          .withTransaction(manager)
+          .emit(PaymentCollectionService.Events.REFUND_FAILED, {
+            ...payment,
+            error: refundError,
+          })
+
+        throw new MedusaError(
+          MedusaError.Types.UNEXPECTED_STATE,
+          `Failed to refund Payment ${payment.id}`
+        )
+      }
+
+      await this.eventBusService_
+        .withTransaction(manager)
+        .emit(PaymentCollectionService.Events.REFUND_CREATED, refund)
+
+      return refund
+    })
+  }
+
+  async refund(
+    paymentId: string,
+    amount: number,
+    reason: string,
+    note?: string
+  ): Promise<Refund> {
+    const manager = this.transactionManager_ ?? this.manager_
+    const paymentCollectionRepository = manager.getCustomRepository(
+      this.paymentCollectionRepository_
+    )
+
+    const payCol =
+      await paymentCollectionRepository.getPaymentCollectionIdByPaymentId(
+        paymentId
+      )
+
+    const payment = await this.paymentProviderService_.retrievePayment(
+      paymentId
+    )
+
+    return await this.refundPayment(payCol, payment, amount, reason, note)
+  }
+
+  async refundAll(
+    paymentCollectionId: string,
+    reason: string,
+    note?: string
+  ): Promise<Refund[]> {
+    const payCol = await this.retrieve(paymentCollectionId, {
+      relations: ["payments"],
+    })
+
+    const allRefunds: Refund[] = []
+    for (const payment of payCol.payments) {
+      const refunded = await this.refundPayment(
+        payCol,
+        payment,
+        payment.amount,
+        reason,
+        note
+      ).catch(() => void 0)
+
+      if (refunded) {
```

Comment:
question: Same as for the captured

Reply:
good point, I believe I will just return the original object in case of failure and users check their status.

Reply:
Should we add a warn of comment it somewhere ?

Reply:
I'm thinking about returning an array of Refunds and another of Payments for the failure.
Then the endpoint will return { refunds: [], failure: [] }